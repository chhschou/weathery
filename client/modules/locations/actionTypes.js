export const REQUEST = 'locations/REQUEST'
export const RECEIVE = 'locations/RECEIVE'
// export const ADD = 'locations/ADD';
// export const DELETE = 'locations/DELETE';
export const EDIT = 'locations/EDIT';
// export const COMPLETE = 'locations/COMPLETE';
// export const COMPLETE_ALL = 'locations/COMPLETE_ALL';;
// export const CLEAR_COMPLETED = 'locations/CLEAR_COMPLETED';
