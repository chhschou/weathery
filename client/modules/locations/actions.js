import request from 'superagent'
import { REQUEST, RECEIVE, EDIT } from 'modules/locations/actionTypes'
import actions from 'actions'
import weather from 'modules/weathers'
import adapter from 'adapters/googleGeocode'

export const receiveLocation = (location) => {
  return { type: RECEIVE, location }
}
export const isUserLocationAvailable = () => navigator.geolocation

export const getUserLocation = () => {
  if (isUserLocationAvailable()) {
    return new Promise((resolve, reject) => {
      navigator.geolocation.getCurrentPosition(resolve, reject)
    })
  }
}

export const requestLocation = () => {
  return { type: REQUEST }
}

export const getLocationViaAddr = (lvl2Addr, country) => {
  return (dispatch) => {
    // dispatch(requestLocation())
  }
}

export function editLocation(location) {
  return { type: types.EDIT, location }
}

export const getLocationViaLatLon = (lat, lng, locationId) => {
  return (dispatch, getState) => {
    dispatch(requestLocation())
    const apiKey = process.env.G_GEOCODE_APIKEY
    const baseUrl = 'https://maps.googleapis.com/maps/api/geocode/json'
    const url = `${baseUrl}?latlng=${lat},${lng}&key=${apiKey}`
    return request.get(url)
      .then((geoCodeRes) => {
        const location = {
          id: locationId,
          ...(adapter.extractLocation(geoCodeRes.body))
        }
        return dispatch(receiveLocation(location))
      })
  }
}