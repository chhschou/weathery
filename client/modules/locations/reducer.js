import * as actionTypes from './actionTypes.js'

const initialState = {
  isFetching: false,
  items: []
}

export default function (state = initialState, action) {

  switch (action.type) {
    case actionTypes.REQUEST:
      state.isFetching = true
      return { ...state }
    case actionTypes.RECEIVE:
      const { items } = state
      const location = action.location
      if (items.length) {
        items[1] = location // reserve 0th for user current location
      } else items.push(location)
      location.id = items.indexOf(location)
      state.isFetching = false
      return { ...state }
    default:
      return state
  }
}
