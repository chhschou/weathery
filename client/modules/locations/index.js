import * as actions from './actions'
// import * as components from './components'
import * as actionTypes from './actionTypes'
import reducer from './reducer'
// import * as selectors from './selectors'

export default { actionTypes, actions, reducer }
