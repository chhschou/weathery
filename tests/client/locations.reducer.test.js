import test from 'ava'
import locations from '../../client/modules/locations'
const { reducer, actionTypes } = locations

test('location reducer has initial state', t => {
  const initialState = [{
    id: 1,
    name: null,
    coordinates: { lat: null, lon: null },
  }]

  t.deepEqual(reducer(undefined, {}), initialState)
})


test('handles RECEIVE action', t => {
  const locations = [
    {
      id: 1, name: 'Wellington', countryName: 'New Zealand', coordinates: {
        lat: -41.29,
        lon: 174.77
      }
    },
    {
      id: 2, name: 'some place', countryName: 'some country', coordinates: {
        lat: -1.0,
        lon: -1.0 
      }
    }
  ]

  const nextState = reducer({}, { type: actionTypes.RECEIVE, locations: locations })

  t.is(nextState.length, 2)
  t.deepEqual(nextState[0], locations[0])
})
