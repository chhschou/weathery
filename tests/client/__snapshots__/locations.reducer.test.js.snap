// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles RECEIVE action 1`] = `
Object {
  "coords": Object {
    "elev": "7.9",
    "lat": "-41.29000092",
    "lng": "174.77999878",
  },
  "displayCity": "Wellington",
  "displayFull": "Wellington, New Zealand",
  "id": 0,
  "observeCity": "Taranaki Street Wharf",
  "observeFull": "Taranaki Street Wharf, Wellington",
  "timeStamp": "1504869815",
}
`;

exports[`handles update location with specific id 1`] = `
Object {
  "coords": Object {
    "elev": "7.9",
    "lat": "-41.29000092",
    "lng": "174.77999878",
  },
  "displayCity": "Wellington",
  "displayFull": "Wellington, New Zealand",
  "id": 5,
  "observeCity": "Taranaki Street Wharf",
  "observeFull": "Taranaki Street Wharf, Wellington",
  "timeStamp": "1504869815",
}
`;
